include/range/v3/range/concepts.hpp

```diff
@@ -316,6 +316,7 @@ namespace ranges
 
     namespace cpp20
     {
+        using ranges::borrowed_range;
```

Comment:
Should I add a deprecation message for `ranges::cpp20::borrowed_range` ?

Reply:
Yep.

Reply:
Can you give me some pointers? I tried to look up the standard and I couldn't find a way to add  attributes to concepts.

gcc complains:

```c++
template<typename T>  concept [[deprecated("HEllo")]] True  = true;

static_assert(True<int>);
```

```
<source>:1:47: error: expected unqualified-id

template<typename T>  [[deprecated("HEllo")]] concept True  = true;
```

I guess that this is not possible. I don't know if a deprecated namespace would work?

Reply:
@ericniebler I figured out two styles:

```c++
template <typename R>
RANGES_DEPRECATED("Please use ranges::borrowed_range instead.")
RANGES_INLINE_VAR constexpr bool safe_range = borrowed_range<R>;
```

* + It will work for all major compilers
* - safe_range is not a concept any more and might make problems in overload resolution; on the other hand you get a BIG use this concept instead message
* - might change semantic

```c++
namespace detail
{
template <typename R>
RANGES_DEPRECATED("Please use ranges::borrowed_range instead.")
RANGES_INLINE_VAR constexpr bool safe_range_is_deprecated = true;
}

template <typename R>
CPP_concept_bool safe_range = borrowed_range<R> && detail::safe_range_is_deprecated<R>;
```

* - It will **NOT** work for all major compilers (except for >= clang-10 with `-std=c++2a`); This is also okay, because normally you compile with at least a second compiler.
* + safe_range is a concept like before
* + does not change semantic

https://godbolt.org/z/CMCoNC (Red is good, because it triggered a deprecation message)

Any suggestions?!

Reply:
(I used the second style)

Reply:
I think I prefer the first style.